### PR TITLE
Make `LazyReference` thread-safe

### PR DESCRIPTION
**Pull request by wyrzyk**

### CODE REVIEW

**Comment by wyrzyk** [comment]
Zrzut ekranu 2021-04-9 o 10 56 47

File: src/main/java/com/atlassian/db/replica/internal/LazyReference.java
Comment (wyrzyk) [inline src/main/java/com/atlassian/db/replica/internal/LazyReference.java:9]: - [ ]  hide the change with `compatibleWithPreviousVersion`.

**Comment by wyrzyk** [comment]
The connection leak can occur when we share a connection between two threads and query the database simultaneously. Postgres do not support this, so we shouldn't be affected. Anyway, it's dangerous to use non-thread-safe Lazy reference, and I'd like to fix it.

File: src/test/java/com/atlassian/db/replica/internal/LazyReferenceTest.java
Comment (dagguh) [inline src/test/java/com/atlassian/db/replica/internal/LazyReferenceTest.java:27]: Really nice test

File: src/test/java/com/atlassian/db/replica/internal/LazyReferenceTest.java
Comment (dagguh) [inline src/test/java/com/atlassian/db/replica/internal/LazyReferenceTest.java:51]: We can get a cleaner synchronization code if we turn the `Runnable` into a `Callable`
```suggestion
    private int fetchReference(
        LazyReference<Integer> reference,
        CountDownLatch start,
        CountDownLatch end
    ) throws InterruptedException {
        start.countDown();
        start.await();
        int value = reference.get();
        end.countDown();
        return value;
    }
```
If we don't want to rely on `Integer`, then we can turn `void` into `Void` and we'll be good too.

File: src/main/java/com/atlassian/db/replica/internal/LazyReference.java
Comment (dagguh) [inline src/main/java/com/atlassian/db/replica/internal/LazyReference.java:20]: This CAS should have worked earlier too. If it returns `false`, then another thread was able to `create`, so we're good.
I ran `git checkout upstream/master -- src/main` and the test is still green.

- [ ] make sure the change is needed

**Review by dagguh — CHANGES_REQUESTED** [review]

File: src/main/java/com/atlassian/db/replica/internal/LazyReference.java
Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/internal/LazyReference.java:20]: It's strange. The CAS should not work as create call is not protected by CAS. Also, the test was red for me, and I call it multiple times. The test also failed on CI (see the screenshot).
1. CAS will help us not to set the reference twice.
2. CAS will not protect from allocating a connection.

Comment (dagguh) [reply src/main/java/com/atlassian/db/replica/internal/LazyReference.java:20]: You're right. Maybe I was lucky

Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/internal/LazyReference.java:20]: It's hard to write a concurrency test to reproduce a race condition that is not flaky and doesn't run forever 😢 .

**Review by dagguh — APPROVED** [review]

File: src/main/java/com/atlassian/db/replica/internal/LazyReference.java
Comment (dagguh) [reply src/main/java/com/atlassian/db/replica/internal/LazyReference.java:20]: No worries 🙂 
It's still great that it reproduced a few times.

File: src/main/java/com/atlassian/db/replica/internal/LazyReference.java
Comment (dagguh) [reply src/main/java/com/atlassian/db/replica/internal/LazyReference.java:9]: Yep, let's hide it

**Review by dagguh — CHANGES_REQUESTED** [review]

File: src/test/java/com/atlassian/db/replica/internal/LazyReferenceTest.java
Comment (afaruga-atlassian) [inline src/test/java/com/atlassian/db/replica/internal/LazyReferenceTest.java:27]: Good test

**Comment by afaruga-atlassian** [comment]
Once hide I'm happy to approve.

File: src/main/java/com/atlassian/db/replica/internal/LazyReference.java
Comment (dwatl) [reply src/main/java/com/atlassian/db/replica/internal/LazyReference.java:10]: @wyrzyk I think it's done, right?

**Review by dwatl — APPROVED** [review]

File: src/main/java/com/atlassian/db/replica/internal/LazyReference.java
Comment (afaruga-atlassian) [inline src/main/java/com/atlassian/db/replica/internal/LazyReference.java:59]: So you pivot from `atomicReference` to lock with `object`

**Review by dagguh — APPROVED** [review]
I cannot review this anymore, so I'll just unblock you

**Review by afaruga-atlassian — APPROVED** [review]

**Comment by afaruga-atlassian** [comment]
>  cannot review this anymore, so I'll just unblock you

👍 
Damn, I'm reviewing merged stuff. :trollface:

**Comment by wyrzyk** [comment]
@afaruga-atlassian There's nothing wrong with reviewing merged stuff.

**Pull request merged**
